File: src/prefs.js
//preferences keys
var REFRESH_TIME = "refresh_time";
var USE_DIRECT_LINK = "use_direct_link";
var ONLY_FAVS = "only_favs";
var ANIMATED_ICON = "animated_icon";
var USE_CONTEXT_MENU = "use_contest_menu";
var NEW_TAB = "new_tab_menu";
var DEBUG_ON = "debug_on";

var PREFS_DEFAULT = new Object();
PREFS_DEFAULT[USE_DIRECT_LINK] = false;
PREFS_DEFAULT[REFRESH_TIME] = 500;
PREFS_DEFAULT[ONLY_FAVS] = false;
PREFS_DEFAULT[ANIMATED_ICON] = false;
PREFS_DEFAULT[USE_CONTEXT_MENU] = false;
PREFS_DEFAULT[NEW_TAB] = true;
PREFS_DEFAULT[DEBUG_ON] = false;

function getPref(name) {
  var value = localStorage[name];
  if (value == null) return PREFS_DEFAULT[name];
  return value;
}

function setPref(name, value) {
  localStorage[name] = value;
}

File: common/src/jsMain/kotlin/flags/prefs/Prefs.kt
package flags.prefs

/**
 * User preferences, shared by the worker, popup and options page. Persisted as
 * a flat record under camelCase keys by [toRecord] / [fromRecord].
 */
data class Prefs(
    val refreshTime: Int = DEFAULT_REFRESH_TIME,
    val getTopics: Boolean = DEFAULT_GET_TOPICS,
    val getMps: Boolean = DEFAULT_GET_MPS,
    val onlyFavs: Boolean = DEFAULT_ONLY_FAVS,
    val useDirectLink: Boolean = DEFAULT_USE_DIRECT_LINK,
    val animatedIcon: Boolean = DEFAULT_ANIMATED_ICON,
    val useContextMenu: Boolean = DEFAULT_USE_CONTEXT_MENU,
    val newTab: Boolean = DEFAULT_NEW_TAB,
    val debugOn: Boolean = DEFAULT_DEBUG_ON,
    val openCat: Boolean = DEFAULT_OPEN_CAT,
    val showCat: Boolean = DEFAULT_SHOW_CAT,
    val maxOpenAll: Int = DEFAULT_MAX_OPEN_ALL,
    /** Theme color; null means "use the active site's default". */
    val bgColor: String? = DEFAULT_BG_COLOR,
    /** Topics the user muted in the popup; hidden from the list. */
    val mutedTopics: List<MutedTopic> = emptyList(),
) {
    /** Defaults, shared by the constructor above and [PrefsMapper.fromRecord]. */
    companion object {
        internal const val DEFAULT_REFRESH_TIME = 500
        internal const val DEFAULT_GET_TOPICS = true
        internal const val DEFAULT_GET_MPS = true
        internal const val DEFAULT_ONLY_FAVS = false
        internal const val DEFAULT_USE_DIRECT_LINK = false
        internal const val DEFAULT_ANIMATED_ICON = true
        internal const val DEFAULT_USE_CONTEXT_MENU = false
        internal const val DEFAULT_NEW_TAB = true
        internal const val DEFAULT_DEBUG_ON = false
        internal const val DEFAULT_OPEN_CAT = true
        internal const val DEFAULT_SHOW_CAT = true
        internal const val DEFAULT_MAX_OPEN_ALL = 10

        /** `null` can't be a `const`, so the nullable default is a plain `val`. */
        internal val DEFAULT_BG_COLOR: String? = null
    }
}
